src/Lean/Elab/MutualInductive.lean

```diff
@@ -1088,16 +1279,64 @@ private def elabInductiveViewsPostprocessing (views : Array InductiveView) (res
             addDocStringOf verso ctor.declName ctor.binders doc
 
   runTermElabM fun _ => Term.withDeclName view0.declName do withRef ref do
-    for view in views do withRef view.declId <| Term.applyAttributesAt view.declName view.modifiers.attrs .afterCompilation
+    for view in views do withRef view.declId <|
+      unless (views.any (·.isCoinductive)) do
+        Term.applyAttributesAt view.declName view.modifiers.attrs .afterCompilation
+
+private def elabInductiveViewsPostprocessingCoinductive (views : Array InductiveView)
+     : CommandElabM Unit := do
+  let views := views.map updateViewRemovingFunctorName
+  let view0 := views[0]!
+  let ref := view0.ref
+  applyComputedFields views -- NOTE: any generated code before this line is invalid
+
+  applyDerivingHandlers views
+  -- Docstrings are added during postprocessing to allow them to have checked references to
+  -- the type and its constructors, but before attributes to enable e.g. `@[inherit_doc X]`
+  runTermElabM fun _ => Term.withDeclName view0.declName do withRef ref do
+    for view in views do
+      withRef view.declId do
+        if let some (doc, verso) := view.docString? then
+          addDocStringOf verso view.declName view.binders doc
+      for ctor in view.ctors do
+        withRef ctor.declId do
+          if let some (doc, verso) := ctor.modifiers.docString? then
+            addDocStringOf verso ctor.declName ctor.binders doc
+
+  runTermElabM fun _ => Term.withDeclName view0.declName do withRef ref do
+    for view in views do withRef view.declId <|
+      unless (views.any (·.isCoinductive)) do
+        Term.applyAttributesAt view.declName view.modifiers.attrs .afterCompilation
+
+def InductiveViewToCoinductiveElab (e : InductiveElabStep1) : CoinductiveElabData where
+  declName := e.view.declName
+  ref := e.view.ref
+  modifiers := e.view.modifiers
+  ctorSyntax := e.view.ctors.map (·.ref)
+  isGreatest := e.view.isCoinductive
 
 def elabInductives (inductives : Array (Modifiers × Syntax)) : CommandElabM Unit := do
-  let (elabs, res) ← runTermElabM fun vars => do
-    let elabs ← inductives.mapM fun (modifiers, stx) => mkInductiveView modifiers stx
-    elabs.forM fun e => checkValidInductiveModifier e.view.modifiers
-    checkNoInductiveNameConflicts elabs
-    let res ← elabInductiveViews vars elabs
-    pure (elabs, res)
-  elabInductiveViewsPostprocessing (elabs.map (·.view)) res
+  let elabs ← runTermElabM fun _ =>
+      inductives.mapM fun (modifiers, stx) => mkInductiveView modifiers stx
+
+  let isCoinductive := elabs.any (·.view.isCoinductive)
+
+  if isCoinductive then
+    runTermElabM fun vars => do
+      let elabs := elabs.map fun e => {e with view := updateViewWithFunctorName e.view}
+      elabs.forM fun e => checkValidInductiveModifier e.view.modifiers
+      checkNoInductiveNameConflicts elabs
```

Comment:
I wonder if this check should come before updating the view. I have pushed a test that maybe relates to that.

Maybe say `let flatElabs :=` to be clear that this is about the flat inductive?